## Coursera Capstone Project

This notebook will be used to present **Coursera Capstone Project**

### WEEK 1 - ASSIGNMENT: *Capstone Project Notebook*

In [567]:
import pandas as pd 
import numpy as np

In [568]:
print ('Hello Capstone Project Course!')

Hello Capstone Project Course!


### WEEK 3 - ASSIGNMENT : *Segmenting and Clustering Neighborhoods in Toronto*

------------------------------------------------------------------------------------------------------------------
**PART 1 - Data Extraction From Wikipedia**

In this part, we will extract the data we need from Wikipedia.  
We will use pandas library to import the needed data from Wikipedia including the postal codes of Toronto.

In [569]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

In [570]:
tables = pd.read_html(url)

There are many tables in the url. So we will choose table 0 which includes the data we want.

In [571]:
df = tables[0]

In [572]:
df = df.rename(columns={'Neighbourhood': 'Neighborhood'})
df.head(10)

Postcode           Borough      Neighborhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
5      M6A        North York  Lawrence Heights
6      M6A        North York    Lawrence Manor
7      M7A      Queen's Park      Not assigned
8      M8A      Not assigned      Not assigned
9      M9A  Downtown Toronto      Queen's Park

We will exclude all the rows that have **Borough** not assigned.

In [573]:
df_clean = df[df.Borough != "Not assigned" ]

In [574]:
df_clean.head(10)

Postcode           Borough      Neighborhood
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M6A        North York  Lawrence Heights
6       M6A        North York    Lawrence Manor
7       M7A      Queen's Park      Not assigned
9       M9A  Downtown Toronto      Queen's Park
10      M1B       Scarborough             Rouge
11      M1B       Scarborough           Malvern
13      M3B        North York   Don Mills North

If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

In [575]:
for index, row in df_clean.iterrows():
    if row['Neighborhood'] == 'Not assigned':
        row['Neighborhood'] = row['Borough']
df_clean.head(11)

Postcode           Borough      Neighborhood
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M6A        North York  Lawrence Heights
6       M6A        North York    Lawrence Manor
7       M7A      Queen's Park      Queen's Park
9       M9A  Downtown Toronto      Queen's Park
10      M1B       Scarborough             Rouge
11      M1B       Scarborough           Malvern
13      M3B        North York   Don Mills North
14      M4B         East York  Woodbine Gardens

We will combine into one row all the neighborhoods with the same pastal code separated with a comma.


In [576]:
df_Toronto = df_clean.groupby(['Postcode','Borough'])['Neighborhood'].apply(', '.join).reset_index()
df_Toronto.head(10)

Postcode      Borough                                     Neighborhood
0      M1B  Scarborough                                   Rouge, Malvern
1      M1C  Scarborough           Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough                Guildwood, Morningside, West Hill
3      M1G  Scarborough                                           Woburn
4      M1H  Scarborough                                        Cedarbrae
5      M1J  Scarborough                              Scarborough Village
6      M1K  Scarborough      East Birchmount Park, Ionview, Kennedy Park
7      M1L  Scarborough                  Clairlea, Golden Mile, Oakridge
8      M1M  Scarborough  Cliffcrest, Cliffside, Scarborough Village West
9      M1N  Scarborough                      Birch Cliff, Cliffside West

In [577]:
df_Toronto.shape

(103, 3)

------------------------------------------------------------------------------------------------------------------
**PART 2 - Assign Latitude & Longitude**

In this part we will assign to each of the Toronto postal code a latitude and a longitude.
To do so we will use the below csv including the data per postal code.

In [578]:
Geospatial_data = pd.read_csv('https://cocl.us/Geospatial_data')

In [579]:
Geospatial_data.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

We proceed with joining the two tables in one.

In [580]:
df_TorGeo = df_Toronto.join(Geospatial_data)

In [581]:
df_TorGeo.head()

Postcode      Borough                            Neighborhood Postal Code  \
0      M1B  Scarborough                          Rouge, Malvern         M1B   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union         M1C   
2      M1E  Scarborough       Guildwood, Morningside, West Hill         M1E   
3      M1G  Scarborough                                  Woburn         M1G   
4      M1H  Scarborough                               Cedarbrae         M1H   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

In [582]:
df_TorGeo = df_TorGeo.drop('Postcode', axis=1)
df_TorGeo = df_TorGeo[['Postal Code', 'Borough', 'Neighborhood', 'Latitude', 'Longitude']]
df_TorGeo.head()

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Rouge, Malvern  43.806686   
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

------------------------------------------------------------------------------------------------------------------
**PART 3 - Exploring and clustering the neighborhoods of Toronto**

#### Step 1 - Installation of libraries.

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [583]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


#### Step 2 - Use geopy library to get the latitude and longitude values of Toronto.

Let's locate the geographical coordinates of Toronto

In [584]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="Toronto")
location = geolocator.geocode(address)
latitude_toronto = location.latitude
longitude_toronto = location.longitude
print('The geographical coordinates of Toronto are {}, {}.'.format(latitude_toronto, longitude_toronto))

The geographical coordinates of Toronto are 43.653963, -79.387207.


#### Step 3 - Create a map of Toronto with neighborhoods superimposed on top.

In [585]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude_toronto, longitude_toronto], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood in zip(df_TorGeo['Latitude'], df_TorGeo['Longitude'], df_TorGeo['Borough'], df_TorGeo['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Downtown Toronto. So let's slice the original dataframe and create a new dataframe of the Downtown Toronto data.

In [586]:
df_DowntownToronto = df_TorGeo[df_TorGeo['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
df_DowntownToronto.head()

Postal Code           Borough                 Neighborhood   Latitude  \
0         M4W  Downtown Toronto                     Rosedale  43.679563   
1         M4X  Downtown Toronto  Cabbagetown, St. James Town  43.667967   
2         M4Y  Downtown Toronto         Church and Wellesley  43.665860   
3         M5A  Downtown Toronto                 Harbourfront  43.654260   
4         M5B  Downtown Toronto     Ryerson, Garden District  43.657162   

   Longitude  
0 -79.377529  
1 -79.367675  
2 -79.383160  
3 -79.360636  
4 -79.378937

Let's get the geographical coordinates of Downtown Toronto.

In [587]:
address = 'Downtown Toronto, ON'

geolocator = Nominatim(user_agent="Toronto")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geographical coordinates of Downtown Toronto are {}, {}.'.format(latitude, longitude))

The geographical coordinates of Downtown Toronto are 43.6563221, -79.3809161.


As we did with the whole of Toronto, let's visualise Downtown Toronto neighborhoods in it.

In [588]:
# create map of Downtown Toronto using latitude and longitude values
map_downtowntrt = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(df_DowntownToronto['Latitude'], df_DowntownToronto['Longitude'], df_DowntownToronto['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_downtowntrt)  
    
map_downtowntrt

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Step 4 - Define Foursquare Credentials and Version

In [589]:
CLIENT_ID = 'CMTVO5RXHJ5EDSW52RLCZP40DCIKXCNP1WEUHNZNU35LIM3E' 
CLIENT_SECRET = 'AXCINYUEHGQ45V4DAHC4OM53IR1SRCTQTERMCZHEES1QDFEN' 
VERSION = '20180605' 

#### STEP 5 - Exproring The First Neighbourhood of Downtown Toronto

Let's explore the first neighborhood in our dataframe and get its coordinates

In [590]:
df_DowntownToronto.loc[0, 'Neighborhood']
neighborhood_latitude = df_DowntownToronto.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df_DowntownToronto.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df_DowntownToronto.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Rosedale are 43.6795626, -79.37752940000001.


Let's explore the top 100 venues that are in Rosedat within a radius of 500 meters using Foursquare API

In [591]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=CMTVO5RXHJ5EDSW52RLCZP40DCIKXCNP1WEUHNZNU35LIM3E&client_secret=AXCINYUEHGQ45V4DAHC4OM53IR1SRCTQTERMCZHEES1QDFEN&v=20180605&ll=43.6795626,-79.37752940000001&radius=500&limit=100'

We use the GET request and examine the results

In [592]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5df66eeefb34b5001bfcc8fc'},
 'response': {'headerLocation': 'Rosedale',
  'headerFullLocation': 'Rosedale, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 4,
  'suggestedBounds': {'ne': {'lat': 43.6840626045, 'lng': -79.37131878274371},
   'sw': {'lat': 43.675062595499995, 'lng': -79.38374001725632}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4aff2d47f964a520743522e3',
       'name': 'Rosedale Park',
       'location': {'address': '38 Scholfield Ave.',
        'crossStreet': 'at Edgar Ave.',
        'lat': 43.68232820227814,
        'lng': -79.37893434347683,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.68232820227814,
          'lng': -79.37893434347683}],
        'distance': 32

All the information is in the **items** key. Before we proceed, let's use the **get_category_type** function.

In [593]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

We proceed by cleaning the json file and structuring it into a *pandas* dataframe.

In [594]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name  categories        lat        lng
0         Rosedale Park  Playground  43.682328 -79.378934
1          Whitney Park        Park  43.682036 -79.373788
2  Alex Murray Parkette        Park  43.678300 -79.382773
3        Milkman's Lane       Trail  43.676352 -79.373842

In [595]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

4 venues were returned by Foursquare.


#### Step 6 - Explore All Neighborhoods in Downtown Toronto

Let's create a function to repeat the same process that we applied for Rosedalevto all the neighborhoods in Downtown Toronto

In [596]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

We write the code to run the above function on each neighborhood and create a new dataframe called **_downtown_toronto_venues_**.

In [597]:
downtown_toronto_venues = getNearbyVenues(names=df_DowntownToronto['Neighborhood'],
                                   latitudes=df_DowntownToronto['Latitude'],
                                   longitudes=df_DowntownToronto['Longitude']
                                  )

Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Queen's Park


In [598]:
print(downtown_toronto_venues.shape)
downtown_toronto_venues.head()

(1279, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                     Rosedale              43.679563              -79.377529   
1                     Rosedale              43.679563              -79.377529   
2                     Rosedale              43.679563              -79.377529   
3                     Rosedale              43.679563              -79.377529   
4  Cabbagetown, St. James Town              43.667967              -79.367675   

                  Venue  Venue Latitude  Venue Longitude Venue Category  
0         Rosedale Park       43.682328       -79.378934     Playground  
1          Whitney Park       43.682036       -79.373788           Park  
2  Alex Murray Parkette       43.678300       -79.382773           Park  
3        Milkman's Lane       43.676352       -79.373842          Trail  
4           Cranberries       43.667843       -79.369407          Diner

We have found 1279 venues spit as below per Neighborhood.

In [599]:
downtown_toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            55   
CN Tower, Bathurst Quay, Island airport, Harbou...                     15   
Cabbagetown, St. James Town                                            43   
Central Bay Street                                                     84   
Chinatown, Grange Park, Kensington Market                              94   
Christie                                                               17   
Church and Wellesley                                                   83   
Commerce Court, Victoria Hotel                                        100   
Design Exchange, Toronto Dominion Centre                              100   
First Canadian Place, Underground city                                100   
Harbord, University of Toronto                                         38   
Harbourfront                                                           50   
Harbourfront East, Toronto Islands, Union Station                     100   
Rosedale                                                                4   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        96   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             55   
CN Tower, Bathurst Quay, Island airport, Harbou...                      15   
Cabbagetown, St. James Town                                             43   
Central Bay Street                                                      84   
Chinatown, Grange Park, Kensington Market                               94   
Christie                                                                17   
Church and Wellesley                                                    83   
Commerce Court, Victoria Hotel                                         100   
Design Exchange, Toronto Dominion Centre                               100   
First Canadian Place, Underground city                                 100   
Harbord, University of Toronto                                          38   
Harbourfront                                                            50   
Harbourfront East, Toronto Islands, Union Station                      100   
Rosedale                                                                 4   
Ryerson, Garden District                                               100   
St. James Town                                                         100   
Stn A PO Boxes 25 The Esplanade                                         96   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                              100             100   
Berczy Park                                            55              55   
CN Tower, Bathurst Quay, Island airport, Harbou...     15              15   
Cabbagetown, St. James Town                            43              43   
Central Bay Street                                     84              84   
Chinatown, Grange Park, Kensington Market              94              94   
Christie                                               17              17   
Church and Wellesley                                   83              83   
Commerce Court, Victoria Hotel                        100             100   
Design Exchange, Toronto Dominion Centre              100             100   
First Canadian Place, Unde

In [600]:
print('There are {} uniques categories.'.format(len(downtown_toronto_venues['Venue Category'].unique())))

There are 205 uniques categories.


#### Step 7 - Analyze All Neighborhoods in Downtown Toronto

In [601]:
# one hot encoding
downtown_toronto_onehot = pd.get_dummies(downtown_toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
downtown_toronto_onehot['Neighborhood'] = downtown_toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [downtown_toronto_onehot.columns[-1]] + list(downtown_toronto_onehot.columns[:-1])
downtown_toronto_onehot = downtown_toronto_onehot[fixed_columns]

downtown_toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0             0         0            0                    0                 0   
1             0         0            0                    0                 0   
2             0         0            0                    0                 0   
3             0         0            0                    0                 0   
4             0         0            0                    0                 0   

   Athletics & Sports  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0                   0          0           0           0       0     0    0   
1                   0          0           0           0       0     0    0   
2                   0          0           0           0       0     0    0   
3                   0          0           0           0       0     0    0   
4                   0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0           0                   0       0              0          0         0   
1           0                   0       0              0          0         0   
2           0                   0       0              0          0         0   
3           0                   0       0              0          0         0   
4           0                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Butcher  Café  Camera Store  Candy Store  \
0             0              0        0     0             0            0   
1             0              0        0     0             0            0   
2             0              0        0     0             0            0   
3             0              0        0     0             0            0   
4             0              0        0     0             0            0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chocolate Shop  \
0                     0            0                   0               0   
1                     0            0                   0       

The new dataframe size is:

In [602]:
downtown_toronto_onehot.shape

(1279, 205)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category.

In [603]:
downtown_toronto_grouped = downtown_toronto_onehot.groupby('Neighborhood').mean().reset_index()
downtown_toronto_grouped.head()

Neighborhood  Yoga Studio  \
0                           Adelaide, King, Richmond     0.000000   
1                                        Berczy Park     0.000000   
2  CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   
3                        Cabbagetown, St. James Town     0.000000   
4                                 Central Bay Street     0.011905   

   Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0                0.0  0.000000            0.000000      0.000000   
1                0.0  0.000000            0.000000      0.000000   
2                0.0  0.066667            0.066667      0.066667   
3                0.0  0.000000            0.000000      0.000000   
4                0.0  0.000000            0.000000      0.000000   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0        0.000000         0.000000          0.000000             0.020000   
1        0.000000         0.000000          0.000000             0.000000   
2        0.133333         0.133333          0.133333             0.000000   
3        0.000000         0.000000          0.000000             0.000000   
4        0.000000         0.000000          0.000000             0.011905   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0           0.0       0.0     0.010000                  0.0              0.03   
1           0.0       0.0     0.018182                  0.0              0.00   
2           0.0       0.0     0.000000                  0.0              0.00   
3           0.0       0.0     0.000000                  0.0              0.00   
4           0.0       0.0     0.000000                  0.0              0.00   

   Athletics & Sports  BBQ Joint  Baby Store  Bagel Shop    Bakery      Bank  \
0                 0.0   0.000000         0.0    0.000000  0.030000  0.000000   
1                 0.0   0.018182         0.0    0.018182  0.036364  0.000000   
2                 0.0   0.000000         0.0    0.000000  0.000000  0.000000   
3                 0.0   0.000000         0.0    0.000000  0.046512  0.023256   
4                 0.0   0.000000         0.0    0.000000  0.011905  0.000000   

        Bar  Baseball Stadium  Basketball Stadium     Beach  Bed & Breakfast  \
0  0.030000               0.0            0.000000  0.000000              0.0   
1  0.000000               0.0            0.018182  0.018182              0.0   
2  0.066667               0.0            0.000000  0.000000              0.0   
3  0.000000               0.0            0.000000  0.000000              0.0   
4  0.023810               0.0            0.000000  0.000000              0.0   

   Beer Bar  Beer Store  Belgian Restaurant    Bistro  Boat or Ferry  \
0  0.000000    0.000000                 0.0  0.000000       0.000000   
1  0.036364    0.000000                 0.0  0.018182       0.000000   
2  0.000000    0.000000                 0.0  0.000000       0.066667   
3  0.000000    0.023256                 0.0  0.000000       0.000000   
4  0.000000    0.000000                 0.0  0.000000       0.000000   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0   0.010000  0.000000                  0.01        0.020000      0.0   
1   0.000000  0.000000                  0.00        0.018182      0.0   
2   0.000000  0.066667                  0.00        0.000000      0.0   
3   0.000000  0.000000                  0.00        0.023256      0.0   
4   0.011905  0.000000                  0.00        0.000000      0.0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place   Butcher      Café  \
0          0.00000      0.01      0.030000           0.01  0.000000  0.050000   
1          0.00000      0.00      0.000000           0.00  0.018182  0.036364   
2          0.00000      0.00      0.000000           0.00  0.000000  0.000000   
3          0.00000      0.00      0.000000           0.00  0.023256  0.046512   
4          0.02381      0.0

In [604]:
downtown_toronto_grouped.shape

(18, 205)

Let's print each neighborhood along with the top 5 most common venues

In [605]:
num_top_venues = 5

for hood in downtown_toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = downtown_toronto_grouped[downtown_toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
              venue  freq
0       Coffee Shop  0.07
1              Café  0.05
2        Steakhouse  0.04
3   Thai Restaurant  0.04
4  Sushi Restaurant  0.03


----Berczy Park----
          venue  freq
0   Coffee Shop  0.09
1  Cocktail Bar  0.04
2      Beer Bar  0.04
3        Bakery  0.04
4          Café  0.04


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
              venue  freq
0    Airport Lounge  0.13
1   Airport Service  0.13
2  Airport Terminal  0.13
3   Harbor / Marina  0.07
4               Bar  0.07


----Cabbagetown, St. James Town----
         venue  freq
0  Coffee Shop  0.09
1   Restaurant  0.07
2          Pub  0.05
3  Pizza Place  0.05
4       Bakery  0.05


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.14
1      Sandwich Place  0.05
2  Italian Restaurant  0.05
3      Ice Cream Shop  0.04
4                Café  0.04


----Chinatown, Grange

Let's put that info into a *pandas* dataframe in descending order.

In [606]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [607]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = downtown_toronto_grouped['Neighborhood']

for ind in np.arange(downtown_toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(downtown_toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2  CN Tower, Bathurst Quay, Island airport, Harbo...       Airport Service   
3                        Cabbagetown, St. James Town           Coffee Shop   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café            Steakhouse       Thai Restaurant   
1           Cheese Shop            Steakhouse          Cocktail Bar   
2      Airport Terminal        Airport Lounge              Boutique   
3            Restaurant                  Café           Pizza Place   
4        Sandwich Place    Italian Restaurant          Burger Joint   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Bar                Bakery           Salad Place   
1                  Café    Seafood Restaurant        Farmers Market   
2   Rental Car Location         Boat or Ferry                   Bar   
3    Italian Restaurant                Bakery           Flower Shop   
4                  Café        Ice Cream Shop  Gym / Fitness Center   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0          Burger Joint            Restaurant       Asian Restaurant  
1                Bakery              Beer Bar               Fountain  
2      Sculpture Garden          Airport Gate     Airport Food Court  
3                   Pub                  Bank              Pet Store  
4           Salad Place             Juice Bar    Japanese Restaurant

#### Step 8 - Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [608]:
# set number of clusters
kclusters = 5

downtown_toronto_grouped_clustering = downtown_toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(downtown_toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 4, 0, 0, 3, 2, 0, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [609]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster_Labels', kmeans.labels_)

df_DowntownToronto_merged = df_DowntownToronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
df_DowntownToronto_merged = df_DowntownToronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
df_DowntownToronto_merged.head() 

Postal Code           Borough                 Neighborhood   Latitude  \
0         M4W  Downtown Toronto                     Rosedale  43.679563   
1         M4X  Downtown Toronto  Cabbagetown, St. James Town  43.667967   
2         M4Y  Downtown Toronto         Church and Wellesley  43.665860   
3         M5A  Downtown Toronto                 Harbourfront  43.654260   
4         M5B  Downtown Toronto     Ryerson, Garden District  43.657162   

   Longitude  Cluster_Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.377529             1.0                  Park            Playground   
1 -79.367675             0.0           Coffee Shop            Restaurant   
2 -79.383160             0.0           Coffee Shop      Sushi Restaurant   
3 -79.360636             0.0           Coffee Shop                  Park   
4 -79.378937             0.0           Coffee Shop        Clothing Store   

  3rd Most Common Venue      4th Most Common Venue 5th Most Common Venue  \
0                 Trail              Deli / Bodega     Electronics Store   
1                  Café                Pizza Place    Italian Restaurant   
2   Japanese Restaurant                    Gay Bar            Restaurant   
3                   Pub                     Bakery    Mexican Restaurant   
4        Cosmetics Shop  Middle Eastern Restaurant                  Café   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0   Dumpling Restaurant            Donut Shop      Doner Restaurant   
1                Bakery           Flower Shop                   Pub   
2                   Pub           Men's Store       Bubble Tea Shop   
3        Breakfast Spot            Restaurant                  Café   
4  Fast Food Restaurant           Pizza Place    Italian Restaurant   

      9th Most Common Venue 10th Most Common Venue  
0                   Dog Run         Discount Store  
1                      Bank              Pet Store  
2  Mediterranean Restaurant              Gastropub  
3                   Theater   Gym / Fitness Center  
4                    Bakery             Restaurant

**important**: No data available for some neighbourhoods. So we drop these rows.

In [610]:
df_DowntownToronto_merged=df_DowntownToronto_merged.dropna()

In [611]:
df_DowntownToronto_merged['Cluster_Labels'] = df_DowntownToronto_merged.Cluster_Labels.astype(int)

Finally, let's visualize the resulting clusters

In [612]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(df_DowntownToronto_merged['Latitude'], df_DowntownToronto_merged['Longitude'], df_DowntownToronto_merged['Neighborhood'], df_DowntownToronto_merged['Cluster_Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Step 9 - Examine Clusters

Based on the defining categories, we could assign a name to each cluster.

#### Cluster 1

In [614]:
df_DowntownToronto_merged.loc[df_DowntownToronto_merged['Cluster_Labels'] == 0, df_DowntownToronto_merged.columns[[1] + list(range(5, df_DowntownToronto_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue  \
1   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0           Coffee Shop   
3   Downtown Toronto               0           Coffee Shop   
4   Downtown Toronto               0           Coffee Shop   
5   Downtown Toronto               0           Coffee Shop   
6   Downtown Toronto               0           Coffee Shop   
7   Downtown Toronto               0           Coffee Shop   
8   Downtown Toronto               0           Coffee Shop   
9   Downtown Toronto               0           Coffee Shop   
10  Downtown Toronto               0           Coffee Shop   
11  Downtown Toronto               0           Coffee Shop   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
1             Restaurant                  Café                Pizza Place   
2       Sushi Restaurant   Japanese Restaurant                    Gay Bar   
3                   Park                   Pub                     Bakery   
4         Clothing Store        Cosmetics Shop  Middle Eastern Restaurant   
5                   Café            Restaurant                      Hotel   
6            Cheese Shop            Steakhouse               Cocktail Bar   
7         Sandwich Place    Italian Restaurant               Burger Joint   
8                   Café            Steakhouse            Thai Restaurant   
9               Aquarium                 Hotel                       Café   
10                  Café                 Hotel                 Restaurant   
11                  Café                 Hotel                 Restaurant   
15            Restaurant                  Café                   Beer Bar   
16                  Café            Restaurant                 Steakhouse   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1     Italian Restaurant                Bakery           Flower Shop   
2             Restaurant                   Pub           Men's Store   
3     Mexican Restaurant        Breakfast Spot            Restaurant   
4                   Café  Fast Food Restaurant           Pizza Place   
5         Clothing Store    Italian Restaurant   American Restaurant   
6                   Café    Seafood Restaurant        Farmers Market   
7                   Café        Ice Cream Shop  Gym / Fitness Center   
8                    Bar                Bakery           Salad Place   
9         Scenic Lookout    Italian Restaurant               Brewery   
10                   Bar   American Restaurant                Bakery   
11   American Restaurant                   Gym    Seafood Restaurant   
15                 Hotel    Seafood Restaurant    Italian Restaurant   
16                 Hotel      Asian Restaurant   American Restaurant   

   8th Most Common Venue     9th Most Common Venue 10th Most Common Venue  
1                    Pub                      Bank              Pet Store  
2        Bubble Tea Shop  Mediterranean Restaurant              Gastropub  
3                   Café                   Theater   Gym / Fitness Center  
4     Italian Restaurant                    Bakery             Restaurant  
5                  Diner            Breakfast Spot               Beer Bar  
6                 Bakery                  Beer Bar               Fountain  
7            Salad Place                 Juice Bar    Japanese Restaurant  
8           Burger Joint                Restaurant       Asian Restaurant  
9    Fried Chicken Joint                Restaurant                  Plaza  
10             Gastropub        Seafood Restaurant          Deli / Bodega  
11            Steakhouse        Italian Restaurant          Deli / Bodega  
15   Japanese Restaurant               Cheese Shop            Art Gallery  
16                   Bar        Seafood Restaurant    Japanese Resta

#### Cluster 2

In [615]:
df_DowntownToronto_merged.loc[df_DowntownToronto_merged['Cluster_Labels'] == 1, df_DowntownToronto_merged.columns[[1] + list(range(5, df_DowntownToronto_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue  \
0  Downtown Toronto               1                  Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Playground                 Trail         Deli / Bodega   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0     Electronics Store   Dumpling Restaurant            Donut Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant               Dog Run         Discount Store

#### Cluster 3

In [616]:
df_DowntownToronto_merged.loc[df_DowntownToronto_merged['Cluster_Labels'] == 2, df_DowntownToronto_merged.columns[[1] + list(range(5, df_DowntownToronto_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue  \
17  Downtown Toronto               2         Grocery Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
17                  Café                  Park           Coffee Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
17            Restaurant    Athletics & Sports    Italian Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
17                 Diner           Candy Store             Baby Store

#### Cluster 4

In [617]:
df_DowntownToronto_merged.loc[df_DowntownToronto_merged['Cluster_Labels'] == 3, df_DowntownToronto_merged.columns[[1] + list(range(5, df_DowntownToronto_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue  \
12  Downtown Toronto               3                  Café   
13  Downtown Toronto               3                  Café   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
12   Japanese Restaurant    Italian Restaurant        Sandwich Place   
13                   Bar    Mexican Restaurant   Dumpling Restaurant   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
12              Bookstore            Restaurant                Bakery   
13  Vietnamese Restaurant                Bakery           Coffee Shop   

   8th Most Common Venue          9th Most Common Venue  \
12               Theater                            Bar   
13    Chinese Restaurant  Vegetarian / Vegan Restaurant   

     10th Most Common Venue  
12       Chinese Restaurant  
13  Comfort Food Restaurant

#### Cluster 5

In [618]:
df_DowntownToronto_merged.loc[df_DowntownToronto_merged['Cluster_Labels'] == 4, df_DowntownToronto_merged.columns[[1] + list(range(5, df_DowntownToronto_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue  \
14  Downtown Toronto               4       Airport Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14      Airport Terminal        Airport Lounge              Boutique   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14   Rental Car Location         Boat or Ferry                   Bar   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14      Sculpture Garden          Airport Gate     Airport Food Court

Thanks!